From de0829b4fb2ac51c25b3d43d2143880cef3c80d6 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:28:45 -0400
Subject: [alsa] hda: Inform the unexpectedly ignored pins by auto-parser

Message-id: <140972823573-244-patchset-jkysela@redhat.com>
Patchwork-id: 90944
O-Subject: [RHEL 7.1 PATCH 243/442] ALSA: hda - Inform the unexpectedly ignored pins by auto-parser
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

The auto-parser may ignore some pins that could be valid when they
don't match with the assumption or if there are way too many pins
assigned to the same output type.  So far, such a pin has been
silently ignored, but it's better to leave a message, which would help
for debugging and understanding the problem.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 9b7564a64999597844513604df4a206fa4da3b69)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/hda_auto_parser.c b/sound/pci/hda/hda_auto_parser.c
index 8de7cfa..90d2fda 100644
--- a/sound/pci/hda/hda_auto_parser.c
+++ b/sound/pci/hda/hda_auto_parser.c
@@ -227,10 +227,18 @@ int snd_hda_parse_pin_defcfg(struct hda_codec *codec,
     continue;
    if (!assoc_line_out)
     assoc_line_out = assoc;
-   else if (assoc_line_out != assoc)
+   else if (assoc_line_out != assoc) {
+    codec_info(codec,
+        "ignore pin 0x%x with mismatching assoc# 0x%x vs 0x%x\n",
+        nid, assoc, assoc_line_out);
     continue;
-   if (cfg->line_outs >= ARRAY_SIZE(cfg->line_out_pins))
+   }
+   if (cfg->line_outs >= ARRAY_SIZE(cfg->line_out_pins)) {
+    codec_info(codec,
+        "ignore pin 0x%x, too many assigned pins\n",
+        nid);
     continue;
+   }
    line_out[cfg->line_outs].pin = nid;
    line_out[cfg->line_outs].seq = seq;
    cfg->line_outs++;
@@ -238,8 +246,12 @@ int snd_hda_parse_pin_defcfg(struct hda_codec *codec,
   case AC_JACK_SPEAKER:
    seq = get_defcfg_sequence(def_conf);
    assoc = get_defcfg_association(def_conf);
-   if (cfg->speaker_outs >= ARRAY_SIZE(cfg->speaker_pins))
+   if (cfg->speaker_outs >= ARRAY_SIZE(cfg->speaker_pins)) {
+    codec_info(codec,
+        "ignore pin 0x%x, too many assigned pins\n",
+        nid);
     continue;
+   }
    speaker_out[cfg->speaker_outs].pin = nid;
    speaker_out[cfg->speaker_outs].seq = (assoc << 4) | seq;
    cfg->speaker_outs++;
@@ -247,8 +259,12 @@ int snd_hda_parse_pin_defcfg(struct hda_codec *codec,
   case AC_JACK_HP_OUT:
    seq = get_defcfg_sequence(def_conf);
    assoc = get_defcfg_association(def_conf);
-   if (cfg->hp_outs >= ARRAY_SIZE(cfg->hp_pins))
+   if (cfg->hp_outs >= ARRAY_SIZE(cfg->hp_pins)) {
+    codec_info(codec,
+        "ignore pin 0x%x, too many assigned pins\n",
+        nid);
     continue;
+   }
    hp_out[cfg->hp_outs].pin = nid;
    hp_out[cfg->hp_outs].seq = (assoc << 4) | seq;
    cfg->hp_outs++;
@@ -267,8 +283,12 @@ int snd_hda_parse_pin_defcfg(struct hda_codec *codec,
    break;
   case AC_JACK_SPDIF_OUT:
   case AC_JACK_DIG_OTHER_OUT:
-   if (cfg->dig_outs >= ARRAY_SIZE(cfg->dig_out_pins))
+   if (cfg->dig_outs >= ARRAY_SIZE(cfg->dig_out_pins)) {
+    codec_info(codec,
+        "ignore pin 0x%x, too many assigned pins\n",
+        nid);
     continue;
+   }
    cfg->dig_out_pins[cfg->dig_outs] = nid;
    cfg->dig_out_type[cfg->dig_outs] =
     (loc == AC_JACK_LOC_HDMI) ?
-- 
1.7.1